src/main/java/com/arpnetworking/metrics/impl/ApacheHttpSink.java

```diff
@@ -50,98 +54,116 @@
                 .build();
-        _uri = builder._uri;
-        _bufferSize = builder._bufferSize;
-        _maxBatchSize = builder._maxBatchSize;
-        _sleepMillis = builder._emptyQueueSleepMillis;
+
         for (int i = 0; i < builder._parallelism; i++) {
             executor.execute(new HttpDispatch());
         }
+
+        Runtime.getRuntime().addShutdownHook(new ShutdownHookThread(this._isRunning));
     }
 
-    private final CloseableHttpClient _httpClient;
-    private final String _uri;
+    private final URI _uri;
     private final int _bufferSize;
     private final int _maxBatchSize;
-    private final int _sleepMillis;
+    private final Duration _emptyQueueInterval;
+    private final CloseableHttpClient _httpClient;
     private final ConcurrentLinkedDeque<Event> _events = new ConcurrentLinkedDeque<>();
     private final AtomicInteger _eventsCount = new AtomicInteger(0);
-//    private final AtomicBoolean _run = new AtomicBoolean(true);
+    private final AtomicBoolean _isRunning = new AtomicBoolean(true);
 
+    private static final Header CONTENT_TYPE_HEADER = new BasicHeader(HttpHeaders.CONTENT_TYPE, "application/octet-stream");
     private static final org.slf4j.Logger LOGGER = LoggerFactory.getLogger(ApacheHttpSink.class);
 
     private final class HttpDispatch implements Runnable {
+
         @Override
         public void run() {
-            ClientV1.RecordSet.Builder requestBuilder = ClientV1.RecordSet.newBuilder();
-            while (true) {
+            while (_isRunning.get()) {
                 // Collect a set of events to send
                 // NOTE: This also builds the serialization types in order to spend more time
                 // allowing more records to arrive in the batch
                 int collected = 0;
 
-                Event event;
+                final ClientV1.RecordSet.Builder requestBuilder = ClientV1.RecordSet.newBuilder();
                 do {
-                    event = _events.pollFirst();
+                    final @Nullable Event event = _events.pollFirst();
                     if (event == null) {
                         break;
                     }
 
                     _eventsCount.decrementAndGet();
                     collected++;
+
                     requestBuilder.addRecords(serializeEvent(event));
                 } while (collected < _maxBatchSize);
 
                 if (collected > 0) {
-                    final ClientV1.RecordSet recordSet = requestBuilder.build();
-                    dispatchRequest(recordSet);
-
-                    // Prepare the next request builder
-                    requestBuilder = ClientV1.RecordSet.newBuilder();
+                    dispatchRequest(requestBuilder.build());
                 }
 
                 // Sleep if we didn't max out the batch size
-                if (event == null) {
-                    uninterruptableSleep(_sleepMillis);
+                if (collected < _maxBatchSize) {
```

Comment:
The comment was wrong, but this dramatically changes the behavior.  I'm guessing that you also had to modify tests to get this to pass?